### ISSUE TITLE
.NET Microservices eBook is being updated to .NET Core 2.1 "wave" of technologies!

### ISSUE THREAD

**Issue by CESARDELATORRE** [dotnet-architecture/News#17]
The .NET Microservices eBook is being updated to .NET Core 2.1 "wave" of technologies!

Main subjects/chapters updated:

**- HttpClientFactory + Polly for resilient Http communication in .NET Core 2.1
- API Gateways and implementation with Ocelot
- Minor updates related to .NET core 2.1 versions, packages and Docker images**

Download from: [https://aka.ms/microservicesebook](https://aka.ms/microservicesebook)
![image](https://user-images.githubusercontent.com/1712635/41567367-ae6ebe58-7314-11e8-92e8-439698e8919a.png)

Note that this is still an on-going effort, hence the version "v2.1.**01**".

In the upcoming weeks it'll be updated in a few more topics and the version will be updated to the next minor versions, such as "v2.1.**01**", "v2.1.**02**", etc.

But we think the community appreciates to have sooner updates in the guidance instead of a single big update coming later.

**Comment by Sn3b** [comment]
Great news! Would you please also update the ePub and mobi versions too?

**Comment by ayayalar** [comment]
+1
		
		

		Arif
	




On Mon, Jun 18, 2018 at 11:08 PM -0700, "Sebastien de Salvador" <notifications@github.com> wrote:












Great news! Would you please also update the ePub and mobi versions too?

**Comment by CESARDELATORRE** [comment]
@sn3b @ayayalar - The eReaders (mobi and epub) versions will be updated in the upcoming weeks, ok? :)

**Comment by CESARDELATORRE** [comment]
@maulikk2000 - Is there any reason why you cannot use HttpClientFactory in .NET Core 2.1 for your scenario? - That’s the recommended way to use HttpClient so it re-uses the pool of handlers plus it is easi r to use policies.
If possible, use HttpClientFactory, either with typed clients, named clients or even directly.
Take a look to the section about HttpClientFactory in the eBook and also to the references I put about HttpClientFactory.
If you had issues, tell me so we’ll look for further possibilities, ok? :)

**Comment by ayayalar** [comment]
Thank you Cesar!!!
		

		Arif
	





On Tue, Jun 19, 2018 at 7:41 PM -0700, "Cesar De la Torre" <notifications@github.com> wrote:












@Sn3b @ayayalar - The eReaders (mobi and epub) versions will be updated in the upcoming weeks, ok? :)

**Comment by maulikk2000** [comment]
Hi Ceasar,

thanks for the prompt reply.

Actually I am using HttpClientFactory  in my API ( source code at
https://github.com/maulikk2000/openhack )  . But my problem is  *" Internal
Server Error The remote certificate is invalid according to the validation
procedure"*   and to overcome in dev I need to use  ServerCert
ificateCustomValidationCallback  which returns true and that method is on
HttpClientHandler.

*Excerpt from my startup.cs*

services.AddHttpClient("openhackAPI", client =>
            {

                var uri = "http://127.0.0.1:8080/";
                client.BaseAddress = new Uri(uri);
                client.DefaultRequestHeaders.Add("Accept",
"application/json");
                client.DefaultRequestHeaders.Authorization = new
AuthenticationHeaderValue("Bearer", bearerToken);
            });

*Controller*

public async Task<IActionResult> Get()
        {
            var client = _httpClientFactory.CreateClient("openhackAPI");

            var data = await
client.GetAsync("api/v1/namespaces/default/services");
            return Ok(data);

        }


In .net core 2.0 I could write like (from
https://stackoverflow.com/questions/38138952/bypass-invalid-ssl-certificate-in-net-core
)

using (var httpClientHandler = new HttpClientHandler()){
   httpClientHandler.ServerCertificateCustomValidationCallback =
(message, cert, chain, errors) => { return true; };
   using (var client = new HttpClient(httpClientHandler))
   {
       // Make your request...
   }}

But I couldnt find a similar way in HttpClientFactory to do it.


Basically my question is, how do I use  ServerCert
ificateCustomValidationCallback   with HttpClientFactory?

thanks,
Maulik

**Comment by onizet** [comment]
```c#
var clientBuilder = services.AddHttpClient("openhackAPI", client => { ... });
clientBuilder.ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler() {
   ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
});
```


Btw, I'm curious about your Bearer token and its lifetime (unless this is an offline_token).
I forward the call to another internal server by providing the received bearer token, with the use of a MessageHandler:
```c#
var clientBuilder = services.AddHttpClient("openhackAPI", client => { ... })
      .AddHttpMessageHandler<BearerTokenHandler>();

/// <summary>
/// Provides the Bearer authentication header from the current caller.
/// </summary>
sealed class BearerTokenHandler : DelegatingHandler
{
     private readonly IHttpContextAccessor httpContextAccessor;

     public BearerTokenHandler(IHttpContextAccessor httpContextAccessor)
     {
         this.httpContextAccessor = httpContextAccessor;
     }

     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
         CancellationToken cancellationToken)
     {
         var context = httpContextAccessor.HttpContext;
         if (context.Request.Headers.TryGetValue("Authorization", out StringValues bearer))
             request.Headers.Authorization = AuthenticationHeaderValue.Parse(bearer.ToString());

         return await base.SendAsync(request, cancellationToken);
     }
}
```

**Comment by sashcatk** [comment]
Re: [dotnet-architecture/News] .NET Microservices eBook is being updated to .NET Core 2.1 "wave" of technologies! (#17)


What the fuck ?

Вы писали 20 июня 2018 г., 13:42:20:





var clientBuilder = services.AddHttpClient("openhackAPI", client => { ... });
clientBuilder.ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler() {
   ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
});
Btw, I'm curious about your Bearer token and its lifetime (unless this is an offline_token).
I forward the call to another internal server by providing the received bearer token, with the use of a MessageHandler:
var clientBuilder = services.AddHttpClient("openhackAPI", client => { ... })

      .AddHttpMessageHandler<BearerTokenHandler>();



/// <summary>

/// Provides the Bearer authentication header from the current caller.

/// </summary>

sealed class BearerTokenHandler : DelegatingHandler

{

     private readonly IHttpContextAccessor httpContextAccessor;



     public BearerTokenHandler(IHttpContextAccessor httpContextAccessor)

     {

         this.httpContextAccessor = httpContextAccessor;

     }



     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,

         CancellationToken cancellationToken)

     {

         var context = httpContextAccessor.HttpContext;

         if (context.Request.Headers.TryGetValue("Authorization", out StringValues bearer))

             request.Headers.Authorization = AuthenticationHeaderValue.Parse(bearer.ToString());



         return await base.SendAsync(request, cancellationToken);

     }

}
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.





--
Зайцев Александр                        mailto:sasha@tk-chel.ru

**Comment by maulikk2000** [comment]
awesome..

thanks for the help. It works like a charm. It's the thing what I was after.

thanks heaps,
Maulik

**Comment by CESARDELATORRE** [comment]
@onizet @maulikk2000 - Right, comment from @onizet about setting the PrimaryHandler is how you must configure HttpClientHandler. 

There's some more info here: 
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-2.1#configure-the-httpmessagehandler

I told the ASP.NET team that we should have some further documentation bout this topic. 👍
